src/threadpool.c

```diff
@@ -29,350 +29,277 @@
+  struct uv__work* wreq;
+  int assigned;
+  int already_completed;
+  int still_on_queue;
+  int can_cancel;
+
+  assert(executor == &default_executor);
+  /* Make sure we are initialized internally. */
+  uv_once(&start_workers_once, start_workers);
+
+  fields = (struct default_executor_fields *) executor->data;
+	assert(fields);
+  wreq = (struct uv__work *) req->reserved[0];
+	assert(wreq);
+
+  uv_mutex_lock(&fields->mutex);
+
+  /* Check if we can cancel it. Determine what state req is in. */
+  assigned = QUEUE_EMPTY(&wreq->wq);
+  already_completed = (wreq->work == NULL);
+  still_on_queue = !assigned && !already_completed;
+  
+  LOG_3("assigned %d already_completed %d still_on_queue\n", assigned, already_completed, still_on_queue); 
+
+  can_cancel = still_on_queue;
+  if (can_cancel)
+    QUEUE_REMOVE(&wreq->wq);
+
+  uv_mutex_unlock(&fields->mutex);
+
+	LOG_1("uv__default_executor_cancel: can_cancel %d\n", can_cancel);
+  if (can_cancel) {
+    /* We are now done with req. Notify libuv.
+     * Note that event loop can't tell we were cancelled yet,
+     * not until later in uv_cancel, but event loop won't go until
+     * uv_cancel returns b/c libuv is single-threaded. */
+    executor->done(req);
+    return 0;
+  }
+  else {
+    /* Failed to cancel.
+     * Work is either already done or is still to be executed.
+     * Either way we need not call done here. */
+    return UV_EBUSY;
+  }
 }
 
-
-int uv_queue_work(uv_loop_t* loop,
-                  uv_work_t* req,
-                  uv_work_cb work_cb,
-                  uv_after_work_cb after_work_cb) {
-  if (work_cb == NULL)
-    return UV_EINVAL;
-
-  uv__req_init(loop, req, UV_WORK);
-  req->loop = loop;
-  req->work_cb = work_cb;
-  req->after_work_cb = after_work_cb;
-  uv__work_submit(loop,
-                  &req->work_req,
-                  UV__WORK_CPU,
-                  uv__queue_work,
-                  uv__queue_done);
-  return 0;
+void uv__default_executor_init(void) {
+  /* TODO Behavior on fork? */
+  default_executor.data = &_fields;
+  default_executor.submit = uv__default_executor_submit;
+  default_executor.cancel = uv__default_executor_cancel;
 }
 
-
-int uv_cancel(uv_req_t* req) {
-  struct uv__work* wreq;
-  uv_loop_t* loop;
-
-  switch (req->type) {
-  case UV_FS:
-    loop =  ((uv_fs_t*) req)->loop;
-    wreq = &((uv_fs_t*) req)->work_req;
-    break;
-  case UV_GETADDRINFO:
-    loop =  ((uv_getaddrinfo_t*) req)->loop;
-    wreq = &((uv_getaddrinfo_t*) req)->work_req;
-    break;
-  case UV_GETNAMEINFO:
-    loop = ((uv_getnameinfo_t*) req)->loop;
-    wreq = &((uv_getnameinfo_t*) req)->work_req;
-    break;
-  case UV_WORK:
-    loop =  ((uv_work_t*) req)->loop;
-    wreq = &((uv_work_t*) req)->work_req;
-    break;
-  default:
-    return UV_EINVAL;
-  }
-
-  return uv__work_cancel(loop, req, wreq);
+uv_executor_t * uv__default_executor(void) {
+  uv_once(&init_default_executor_once, uv__default_executor_init);
+  return &default_executor;
 }
```

Comment:
I will admit that I don't really the see the value of grouping everything in that `_fields` struct when it's still all global state.

Reply:
It made a bit more sense when the executor API had `init` and `destroy` callbacks. I think I will revert `_fields` now that those are gone.